### PR TITLE
[Microsoft] Added ability to retrieve tenant information when using multi-tenant auth

### PR DESCRIPTION
**Pull request by allw1994**
Added ability to retrieve tenant information when using multi-tenant auth to allow for tenant filtering for any apps that may wish to form allow/blocklists or who may want to only allow access to "internal" tenants.

### CODE REVIEW

File: src/Microsoft/Provider.php [hunk 1]
```diff
@@ -58,20 +64,39 @@ protected function getTokenUrl()
      */
     protected function getUserByToken($token)
     {
-        $response = $this->getHttpClient()->get(
+        $response_user = $this->getHttpClient()->get(
```
Comment (atymic) [inline src/Microsoft/Provider.php:67]:
```suggestion
        $responseUser = $this->getHttpClient()->get(
```

Camelcase please

Comment (atymic) [reply src/Microsoft/Provider.php:67]: (and in other spots)

File: src/Microsoft/Provider.php [hunk 2]
```diff
@@ -58,20 +64,39 @@ protected function getTokenUrl()
...
-                    '$select' => implode(',', array_merge(self::DEFAULT_FIELDS, ($this->config['fields'] ?: []))),
+                    '$select' => implode(',', array_merge(self::DEFAULT_FIELDS_USER, ($this->config['user_fields'] ?: []))),
                 ],
             ]
         );
 
-        return json_decode((string) $response->getBody(), true);
+        $formatted_response = json_decode((string) $response_user->getBody(), true);
+
+        if ($this->getConfig('tenant', 'common') === 'common') {
```
Comment (atymic) [inline src/Microsoft/Provider.php:82]: should we not make this opt in, rather than opt out?

**Review by atymic — CHANGES_REQUESTED** [review]

File: src/Microsoft/Provider.php (hunk 2 above)
Comment (allw1994) [reply src/Microsoft/Provider.php]: Have included in further changes.

**Comment by allw1994** [comment]
Does anything else still need to be tweaked?

File: src/Microsoft/Provider.php [hunk 3]
```diff
@@ -58,20 +64,39 @@ protected function getTokenUrl()
...
                 ],
                 RequestOptions::QUERY => [
-                    '$select' => implode(',', array_merge(self::DEFAULT_FIELDS, ($this->config['fields'] ?: []))),
+                    '$select' => implode(',', array_merge(self::DEFAULT_FIELDS_USER, ($this->config['fields'] ?: []))),
                 ],
             ]
         );
 
-        return json_decode((string) $response->getBody(), true);
+        $formatted_response = json_decode((string) $responseUser->getBody(), true);
```
Comment (atymic) [inline src/Microsoft/Provider.php:80]: camelcase please

File: src/Microsoft/Provider.php [hunk 4]
```diff
@@ -58,20 +64,39 @@ protected function getTokenUrl()
...
+
+        if (($this->getConfig('tenant', 'common') === 'common') && ($this->getConfig('include_tenant_info', false) === true)) {
```
Comment (atymic) [inline src/Microsoft/Provider.php:82]:
```suggestion
        if ($this->getConfig('tenant', 'common') === 'common' && $this->getConfig('include_tenant_info', false) {
```

File: src/Microsoft/Provider.php [hunk 5]
```diff
@@ -58,20 +64,39 @@ protected function getTokenUrl()
...
+                        'Accept'        => 'application/json',
+                        'Authorization' => 'Bearer '.$token,
+                    ],
+                    RequestOptions::QUERY => [
+                        '$select' => implode(',', array_merge(self::DEFAULT_FIELDS_TENANT, ($this->config['tenant_fields'] ?: []))),
+                    ],
+                ]
+            );
+
+            $formatted_response['tenant'] = json_decode((string) $responseTenant->getBody(), true)['value'][0];
```
Comment (atymic) [inline src/Microsoft/Provider.php:96]: as above. Do we need a null coalesce here to make sure this does't error if the request doesn't have results?
```suggestion
            $formatted_response['tenant'] = json_decode((string) $responseTenant->getBody(), true)['value'][0] 
            ?? null;
```

**Review by atymic — CHANGES_REQUESTED** [review]

File: src/Microsoft/Provider.php [hunk 6]
```diff
@@ -58,20 +64,39 @@ protected function getTokenUrl()
...
+        if ($this->getConfig('tenant', 'common') === 'common' && $this->getConfig('include_tenant_info', false)) {
+            $responseTenant = $this->getHttpClient()->get(
+                'https://graph.microsoft.com/v1.0/organization',
+                [
+                    RequestOptions::HEADERS => [
+                        'Accept'        => 'application/json',
+                        'Authorization' => 'Bearer '.$token,
+                    ],
+                    RequestOptions::QUERY => [
+                        '$select' => implode(',', array_merge(self::DEFAULT_FIELDS_TENANT, ($this->config['tenant_fields'] ?: []))),
```
Comment (atymic) [inline src/Microsoft/Provider.php:91]: Sorry one more 🙈 
```suggestion
                        '$select' => implode(',', array_merge(self::DEFAULT_FIELDS_TENANT, $this->getConfig('tenant_fields'))),
```

File: src/Microsoft/Provider.php [hunk 7]
```diff
@@ -58,20 +64,39 @@ protected function getTokenUrl()
...
+        $responseUser = $this->getHttpClient()->get(
             'https://graph.microsoft.com/v1.0/me',
             [
                 RequestOptions::HEADERS => [
                     'Accept'        => 'application/json',
                     'Authorization' => 'Bearer '.$token,
                 ],
                 RequestOptions::QUERY => [
-                    '$select' => implode(',', array_merge(self::DEFAULT_FIELDS, ($this->config['fields'] ?: []))),
+                    '$select' => implode(',', array_merge(self::DEFAULT_FIELDS_USER, ($this->config['fields'] ?: []))),
```
Comment (atymic) [inline src/Microsoft/Provider.php:75]: And may as well update it here as well

**Comment by allw1994** [comment]
Done, I hope - want to make sure I write it into the docs and I have more to add to this file too so need to square this off first.

**Review by atymic — APPROVED** [review]

**Pull request merged**